wagtail/admin/forms/workflows.py

```diff
@@ -149,6 +152,98 @@ def clean(self):
             )
 
 
+class WorkflowContentTypeForm(forms.Form):
+    class ContentTypeMultipleChoiceField(forms.ModelMultipleChoiceField):
+        def label_from_instance(self, obj):
+            return capfirst(obj.name)
+
+    class CheckboxSelectMultiple(forms.CheckboxSelectMultiple):
+        """Custom CheckboxSelectMultiple widget that renders errors for each content type ID"""
+
+        option_template_name = (
+            "wagtailadmin/workflows/includes/workflow_content_types_checkbox.html"
+        )
+
+        def get_errors_by_id(self, errors):
+            errors_by_id = {}
+            for error in errors.as_data():
+                ct_id = error.params and error.params.get("content_type_id")
+                errors_by_id.setdefault(ct_id, []).append(error)
+            return errors_by_id
+
+        def render_with_errors(
+            self, name, value, attrs=None, renderer=None, errors=None
+        ):
+            context = {
+                **self.get_context(name, value, attrs),
+                "errors_by_id": self.get_errors_by_id(errors),
+            }
+            return self._render(self.template_name, context, renderer)
+
+    content_types = ContentTypeMultipleChoiceField(
+        queryset=ContentType.objects.none(),
+        widget=CheckboxSelectMultiple(),
+        required=False,
+    )
+
+    def __init__(self, *args, workflow=None, **kwargs):
+        self.workflow = workflow
+        if workflow and "initial" not in kwargs:
+            kwargs["initial"] = {"content_types": workflow.workflow_content_types.all()}
+
+        super().__init__(*args, **kwargs)
+
+        workflow_cts = ContentType.objects.get_for_models(
+            *get_workflow_enabled_models()
+        ).values()
+        self.fields["content_types"].queryset = ContentType.objects.filter(
+            pk__in=[ct.pk for ct in workflow_cts]
```

Comment:
I don't love the double query here, though it's probably not worth nitpicking over a single additional query! But you could get it in one by just querying directly on app label and model name for each (combining Q objects to represent the ANDs and ORs correctly)